tools/deploy/helper.go

```diff
@@ -0,0 +1,438 @@
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, 1)
+	)
+	for i := range targets {
+		i := i
+		wg.Add(1)
+		h.workers.Go(func() {
+			defer wg.Done()
+
+			err := h.foreverRetrier.Attempt(func() error {
+				validator := targets[i].Validator
+				if validator == nil {
+					return nil
+				}
+				if err := validator(); err != nil {
+					err = fmt.Errorf("validation error for instance %s: %v", targets[i].Instance.PlacementID, err)
+					return err
+				}
+				return nil
+			})
+			if err == nil {
+				return
+			}
+			select {
+			case errCh <- err:
+			default:
+			}
+		})
+	}
+	wg.Wait()
+	close(errCh)
+	return <-errCh
+}
+
+func (h helper) resign(targets []deploymentTarget) error {
+	var (
+		wg    sync.WaitGroup
+		errCh = make(chan error, 1)
+	)
+	for i := range targets {
+		i := i
+		wg.Add(1)
+		h.workers.Go(func() {
+			defer wg.Done()
+
+			err := h.retrier.Attempt(func() error {
+				instance := targets[i].Instance
+				if err := h.client.Resign(instance.APIEndpoint); err != nil {
+					err = fmt.Errorf("resign error for instance %s: %v", instance.PlacementID, err)
+					return err
+				}
+				return nil
+			})
+			if err == nil {
+				return
+			}
+			select {
+			case errCh <- err:
+			default:
+			}
+		})
+	}
+	wg.Wait()
+	close(errCh)
+	return <-errCh
+}
+
+func (h helper) deploy(targetIDs []string, revision string) error {
+	return h.retrier.Attempt(func() error {
+		return h.mgr.Deploy(targetIDs, revision)
+	})
+}
+
+func (h helper) waitUntilProgressing(targetIDs []string, revision string) error {
+	return h.foreverRetrier.Attempt(func() error {
+		targetInstances, err := h.mgr.Query(targetIDs)
+		if err != nil {
+			return fmt.Errorf("error querying instances: %v", err)
+		}
+
+		for _, di := range targetInstances {
+			if di.IsDeploying() || di.Revision() == revision {
+				return nil
+			}
+		}
+
+		return errNoDeploymentProgress
+	})
+}
+
+func (h helper) allInstanceMetadatas() (instanceMetadatas, error) {
+	placementInstances, err := h.placementInstances()
+	if err != nil {
+		return nil, fmt.Errorf("unable to determine instances from placement: %v", err)
+	}
+	deploymentInstances, err := h.mgr.QueryAll()
+	if err != nil {
+		return nil, fmt.Errorf("unable to query all instances from deployment: %v", err)
+	}
+	metadatas, err := h.computeInstanceMetadatas(placementInstances, deploymentInstances)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compute instance metadatas: %v", err)
+	}
+	return metadatas, nil
+}
+
+// validateInstances validates instances derived from placement against
+// instances derived from deployment, ensuring there are no duplicate instances
+// and the instances derived from two sources match against each other.
+func (h helper) computeInstanceMetadatas(
+	placementInstances []services.PlacementInstance,
+	deploymentInstances []Instance,
+) (instanceMetadatas, error) {
+	if len(placementInstances) != len(deploymentInstances) {
```

Comment:
Would we do deployments on hot spares in the future? In which case they will not be in the placement